srml/contract/src/wasm/runtime.rs

```diff
@@ -628,4 +628,16 @@ define_env!(Env, <E: Ext>,
 
 		Ok(())
 	},
+
+	// Prints utf8 encoded string from the data buffer.
+	// MUST only be called from contracts deployed to `--dev` chains, NOT for use in production chains.
+	// This function may be removed in the future and superseded by "Diagnostic Runtimes":
+	// https://github.com/paritytech/substrate/issues/2082
+	ext_println(ctx, str_ptr: u32, str_len: u32) => {
+		let data = read_sandbox_memory(ctx, str_ptr, str_len)?;
+		if let Ok(utf8) = core::str::from_utf8(&data) {
+			runtime_io::print(utf8);
```

Comment:
Why not use `ext_print_utf8` directly?

Reply:
Can we actually use `ext_print_utf8` here?
We are reading from sandbox memory.
Also `ext_print_utf8` seems to make multiple copies of the data that we can avoid here, don't know if they are avoidable in `ext_print_utf8`.
Tbh performance concerns are negligible for printing stuff.

Reply:
Print uses `ext_print_utf8` anyway under the hood. I don't know where you saw that it makes multiple copies of the data.

Reply:
But am I correct in assuming that this code must compile with and without `std`, which the `Print` trait abstracts for us https://github.com/paritytech/substrate/blob/master/core/sr-io/src/lib.rs?

Reply:
Exactly right. At first I also had a question regarding usage of `ext_print_utf8` but it has dropped as soon as I realized that we have to use `std`. And yeah, we don't need to care about performance or DoS factors as long as this code is guarded with `--dev` flag.